src/mpid/ch4/netmod/ofi/util.c

```diff
@@ -509,12 +509,8 @@ static inline void add_index(MPI_Datatype datatype, int *idx)
 
 void MPIDI_OFI_index_datatypes()
 {
-    static bool needs_init = true;
     int idx = 0;
 
-    if (!needs_init)
-        return;
-
     add_index(MPI_CHAR, &idx);
```

Comment:
We already have the `mpi_dtypes[]` static array in `typeutil.c`, could we use that array instead of listing the types manually again here?

In addition, here we don't check `MPI_DATATYPE_NULL`, which should be a problem, right?

Reply:
Yes, we can probably use that array. I work on another patch. I added a check for `MPI_DATATYPE_NULL` to the `add_index` function in another PR. Need to rebase before pushing again.

Reply:
I'm running into a problem `sizeof` on an the arrays after declaring them extern to use in `ofi/util.c`.
```
src/mpid/ch4/netmod/ofi/util.c:444:27: error: invalid application of 'sizeof' to an incomplete type 'mpi_names_t []'
    for (i = 0; i < sizeof(mpi_dtypes) / sizeof(mpi_dtypes[0]); i++)
```

Reply:
Apparently, you can define constants with `sizeof`, so
```
int mpi_dtypes_count = sizeof(mpi_dtypes) / sizeof(mpi_dtypes[0];
```
as extern global variable should work.

Reply:
Added 2 commits for this.

Reply:
After looking more closely, it appears this change will require a more substantial re-write of the datatype mapping code in `ch4/ofi`. I would like to defer this for now. The dependency cleanup was my primary goal, and this PR achieves it.

Reply:
Additionally, I think a more desirable alternative to exposing these arrays might be to provide APIs to iterate over datatype handles: builtin language types, pair types, and *all* types.